src/graph/heterograph.cc

```diff
@@ -532,31 +535,20 @@ bool HeteroGraph::Load(dmlc::Stream* fs) {
   uint64_t magicNum;
   CHECK(fs->Read(&magicNum)) << "Invalid Magic Number";
   CHECK_EQ(magicNum, kDGLSerialize_HeteroGraph) << "Invalid HeteroGraph Data";
-  auto meta_grptr = new ImmutableGraph(static_cast<COOPtr>(nullptr));
-  CHECK(fs->Read(meta_grptr)) << "Invalid Immutable Graph Data";
-  uint64_t num_relation_graphs;
-  CHECK(fs->Read(&num_relation_graphs)) << "Invalid num of relation graphs";
-  std::vector<HeteroGraphPtr> relgraphs;
-  for (size_t i = 0; i < num_relation_graphs; ++i) {
-    UnitGraph* ugptr = Serializer::EmptyUnitGraph();
-    CHECK(fs->Read(ugptr)) << "Invalid UnitGraph Data";
-    relgraphs.emplace_back(dynamic_cast<BaseHeteroGraph*>(ugptr));
-  }
-  HeteroGraph* hgptr = new HeteroGraph(GraphPtr(meta_grptr), relgraphs);
-  *this = *hgptr;
+  auto meta_imgraph = Serializer::make_shared<ImmutableGraph>();
+  CHECK(fs->Read(&meta_imgraph)) << "Invalid meta graph";
+  meta_graph_ = std::dynamic_pointer_cast<GraphInterface>(meta_imgraph);
```

Comment:
The dynamic cast is not needed for parent class. Just `meta_graph_ = meta_imgraph` should be fine.

Reply:
I'm also thinking here whether should make immutable graph as the default serialization of GraphInterface. (Add save/load function to the `GraphInterface`, which implemented using ImmutableGraph's save/load)

Reply:
You cannot avoid creating an explicit child class shared pointer in `Load`. It is essentially a problem of reflection. I think your solution here is fine.